Add `dependsOn` functionality to `FormValidator`

DanTrap opened the pull request.

DanTrap committed:
Add `dependsOn` functionality to `FormValidator`

aartikov commented on kmm-form-validation/src/commonMain/kotlin/ru/mobileup/kmm_form_validation/validation/form/FormValidator.kt:
```diff
@@ -21,8 +21,24 @@ import ru.mobileup.kmm_form_validation.validation.control.ValidationResult
  */
 class FormValidator(
     val validators: Map<UIControl<*>, ControlValidator<*>>,
+    val validatorDependencies: Map<ControlValidator<*>, Set<UIControl<*>>>,
```
Может, лучше в `ControlValidator` добавить `val dependencies: Set<UIControl<*>>`

aartikov commented on kmm-form-validation/src/commonMain/kotlin/ru/mobileup/kmm_form_validation/validation/form/FormValidator.kt:
```diff
@@ -21,8 +21,24 @@ import ru.mobileup.kmm_form_validation.validation.control.ValidationResult
+        setupDependencyListeners()
+    }
+
+    private fun setupDependencyListeners() {
```
Эту логику лучше перенести в RevalidateOnValueChanged, так как именно эта фича отвечает за ревалидацию при смене значений

aartikov commented on kmm-form-validation/src/commonMain/kotlin/ru/mobileup/kmm_form_validation/validation/form/FormValidator.kt:
```diff
@@ -21,8 +21,24 @@ import ru.mobileup.kmm_form_validation.validation.control.ValidationResult
+        validatorDependencies.forEach { (validator, dependsOn) ->
+            dependsOn.forEach { control ->
+                control.value.onEach {
+                    if (validator.isFilled) validator.validate(true)
```
Посмотри, как в RevalidateOnValueChanged сделано. Там ревалидируется поле только если у него есть ошибка

aartikov commented on sample/sharedSample/src/commonMain/kotlin/ru/mobileup/kmm_form_validation/sharedsample/ui/RealFormComponent.kt:
```diff
@@ -147,7 +147,10 @@ class RealFormComponent(
+        input(
+            inputControl = confirmPasswordInput,
+            dependsOn = setOf(passwordInput)
+        ) {
```
Идея - сделать возможность задавать dependsOn как у input так и у validation (все зависимости будут объединяться в одно множество).

А equalsTo автоматически будет добавлять dependsOn на нужный контрол

DanTrap committed:
- Create `BaseValidatorBuilder` to streamline the creation of validators.
- Add 'dependsOn' property to `ControlValidator` to allow validators to depend on other controls.
